docs/Configuration.md
# Log Courier Configuration

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*

- [Overview](#overview)
- [Reloading](#reloading)
- [Examples](#examples)
- [Field Types](#field-types)
  - [String, Number, Boolean, Array, Dictionary](#string-number-boolean-array-dictionary)
  - [Duration](#duration)
  - [Fileglob](#fileglob)
- [Stream Configuration](#stream-configuration)
  - [`"codecs"`](#codecs)
  - [`"dead time"`](#dead-time)
  - [`"fields"`](#fields)
- [`"general"`](#general)
  - [`"admin enabled"`](#admin-enabled)
  - [`"admin listen address"`](#admin-listen-address)
  - [`"log file"`](#log-file)
  - [`"global fields"`](#global-fields)
  - [`"host"`](#host)
  - [`"log level"`](#log-level)
  - [`"log stdout"`](#log-stdout)
  - [`"log syslog"`](#log-syslog)
  - [`"line buffer bytes"`](#line-buffer-bytes)
  - [`"max line bytes"`](#max-line-bytes)
  - [`"persist directory"`](#persist-directory)
  - [`"prospect interval"`](#prospect-interval)
  - [`"spool max bytes"`](#spool-max-bytes)
  - [`"spool size"`](#spool-size)
  - [`"spool timeout"`](#spool-timeout)
- [`"network"`](#network)
  - [`"max pending payloads"`](#max-pending-payloads)
  - [`"method"`](#method)
  - [`"reconnect"`](#reconnect)
  - [`"rfc 2782 srv"`](#rfc-2782-srv)
  - [`"rfc 2782 service"`](#rfc-2782-service)
  - [`"servers"`](#servers)
  - [`"ssl ca"`](#ssl-ca)
  - [`"ssl certificate"`](#ssl-certificate)
  - [`"ssl key"`](#ssl-key)
  - [`"timeout"`](#timeout)
  - [`"transport"`](#transport)
- [`"files"`](#files)
  - [`"paths"`](#paths)
- [`"includes"`](#includes)
- [`"stdin"`](#stdin)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Overview

The Log Courier configuration is currently stored in standard JSON format with
the exception that comments are allowed.

End-of-line comments start with a pound sign outside of a string, and cause all
characters until the end of the line to be ignored. Block comments start with
a forwarder slash and an asterisk and cause all characters, including new lines,
to be ignored until an asterisk followed by a forwarder slash is encountered.

```
{
    "general": {
        # General configuration here
    },
    "network": {
        # Network configuration here
    }, # (these are end-of-line comments)
    "files": {
        /* File configuration here
        (this is a block comment) */
    }
}
```

## Reloading

Log Courier can reload its configuration without the need for a restart. It will
do this upon receiving the SIGHUP signal. To send this signal, run the following
command replacing 1234 with the Process ID of Log Courier.

    kill -HUP 1234

Log Courier will reopen its own log file if one has been configured, allowing
native log rotation to take place.

Please note that files Log Courier has already started harvesting will continue
to be harvested after the reload with their original configuration; the reload
process will only affect new files. Additionally, harvested log files will not
be reopened. Log rotations are detected automatically. To control when a
harvested log file is closed you can adjust the [`"dead time"`](#dead-time)
option.

In the case of a network configuration change, Log Courier will disconnect and
reconnect as required at the earliest opportunity.

*Configuration reload is not currently available on Windows builds of Log
Courier.*

## Examples

Several configuration examples are available for you perusal.

* [Ship a single log file](examples/example-single.conf)
* [Ship a folder of log files](examples/example-folder.conf)
* [Ship from STDIN](examples/example-stdin.conf)
* [Ship logs with extra field information](examples/example-fields.conf)
* [Multiline log processing](examples/example-multiline.conf)

The configuration is documented in full below.

## Field Types

### String, Number, Boolean, Array, Dictionary

These are JSON types and follow the same rules. Strings within double quotes,
arrays of fields within square brackets separated by commas, and dictionaries
of key value pairs within curly braces and each entry, in the form `"key":
value`, separated by a comma.

### Duration

This can be either a number or a string describing the duration. A number will
always be interpreted in seconds.

* `5` = 5 seconds
* `300` = 5 minutes (which is 300 seconds)
* `"5s"` = 5 seconds
* `"15m"` = 15 minutes

### Fileglob

A fileglob is a string representing a file pattern.

The pattern format used is detailed at
http://golang.org/pkg/path/filepath/#Match and is shown below for reference:

```
term:
    '*'         matches any sequence of non-Separator characters
    '?'         matches any single non-Separator character
    '[' [ '^' ] { character-range } ']'
                            character class (must be non-empty)
    c           matches character c (c != '*', '?', '\\', '[')
    '\\' c      matches character c

character-range:
    c           matches character c (c != '\\', '-', ']')
    '\\' c      matches character c
    lo '-' hi   matches character c for lo <= c <= hi
```

* `"/var/log/*.log"`
* `"/var/log/program/log_????.log"`
* `"/var/log/httpd/access.log"`
* `"/var/log/httpd/access.log.[0-9]"`

## Stream Configuration

Stream Configuration parameters can be specified for file groups within
[`"files"`](#files) and also for [`"stdin"`](#stdin). They customise the log
entries produced by passing, for example, by passing them through a codec and
adding extra fields.

### `"codecs"`

*Codec configuration. Optional. Default: `[ { "name": "plain" } ]`*
*Configuration reload will only affect new or resumed files*

*Depending on how log-courier was built, some codecs may not be available. Run
`log-courier -list-supported` to see the list of codecs available in a specific
build of log-courier.*

The specified codecs will receive the lines read from the log stream and perform
any decoding necessary to generate events. The plain codec does nothing and
simply ships the events unchanged.

When multiple codecs are specified, the first codec will receive events, and the
second codec will receive the output from the first codec. This allows versatile
configurations, such as combining events into multiline events and then
filtering those that aren't required.

All configurations are an array of dictionaries with at least a "name" key.
Additional options can be provided if the specified codec allows.

* `[ { "name": "codec-name" } ]`
* `[ { "name": "codec-name", "option1": "value", "option2": "42" } ]`
* `[ { "name": "first-name" }, { "name": "second-name" } ]`

Aside from "plain", the following codecs are available at this time.

* [Filter](codecs/Filter.md)
* [Multiline](codecs/Multiline.md)

### `"dead time"`

*Duration. Optional. Default: "24h"*
*Configuration reload will only affect new or resumed files*

If a log file has not been modified in this time period, it will be closed and
Log Courier will simply watch it for modifications. If the file is modified it
will be reopened.

If a log file that is being harvested is deleted, it will remain on disk until
Log Courier closes it. Therefore it is important to keep this value sensible to
ensure old log files are not kept open preventing deletion.

### `"fields"`

*Dictionary. Optional*
*Configuration reload will only affect new or resumed files*

Extra fields to attach to events prior to shipping. These can be simple strings,
numbers or even arrays and dictionaries.

Examples:

* `{ "type": "syslog" }`
* `{ "type": "apache", "server_names": [ "example.com", "www.example.com" ] }`
* `{ "type": "program", "program": { "exec": "program.py", "args": [ "--run", "--daemon" ] } }`

## `"general"`

The general configuration affects the general behaviour of Log Courier, such
as where to store its persistence data or how often to scan for the appearence
of new log files.

### `"admin enabled"`

*Boolean. Optional. Default: false*  
*Requires restart*

Enables the administration listener that the `lc-admin` utility can connect to.

### `"admin listen address"`

*String. Optional. Default: tcp:127.0.0.1:1234*

The address the administration listener should listen on in the format
`transport:address`.

Allowed transports are "tcp", "tcp4", "tcp6" (Windows and *nix) and "unix"
(*nix only). For the tcp transports the address format is `host:port`. For the
unix transport the address should specify a filename to use when creating the
unix domain socket. If no transport name is specified, "tcp" is assumed.

Examples:

    127.0.0.1:1234
    tcp:127.0.0.1:1234
    unix:/var/run/log-courier/admin.socket

### `"log file"`

*Filepath. Optional*  
*Requires restart*

A log file to save Log Courier's internal log into. May be used in conjunction with `"log stdout"` and `"log syslog"`.

### `"global fields"`

*Dictionary. Optional*
*Configuration reload will only affect new or resumed files*

Extra fields to attach to events prior to shipping. This is identical in
behaviour to the "`fields`" Stream Configuration and applies globally to the
`"stdin"` section and to all files listed in the `"files"` section.

### `"host"`

*String. Optional. Default: System FQDN.*
*Configuration reload will only affect new or resumed files*

Every event has an automatic field, "host", that contains the current system
FQDN. Using this option allows a custom value to be given to the "host" field
instead of the system FQDN.

### `"log level"`

*String. Optional. Default: "info".  
Available values: "critical", "error", "warning", "notice", "info", "debug"*  
*Requires restart*

The minimum level of detail to produce in Log Courier's internal log.

### `"log stdout"`

*Boolean. Optional. Default: true*  
*Requires restart*

Enables sending of Log Courier's internal log to the console (stdout). May be used in conjunction with `"log syslog"` and `"log file"`.

### `"log syslog"`

*Boolean. Optional. Default: false*  
*Requires restart*

Enables sending of Log Courier's internal log to syslog. May be used in conjunction with `"log stdout"` and `"log file"`.

*This option is ignored by Windows builds.*

### `"line buffer bytes"`

*Number. Optional. Default: 16384*

The size of the line buffer used when reading files.

If `max line bytes` is greater than this value, any lines that exceed this size
will trigger additional memory allocations. This value should be set to a value
just above the 90th percentile (or average) line length.

### `"max line bytes"`

*Number. Optional. Default: 1048576*

The maxmimum line length to process. If a line exceeds this length, it will be
split across multiple events. Each split line will have a "tag" field added
containing the tag "splitline". The final part of the line will not have a "tag"
field added.

If the `fields` configuration already contained a "tags" entry, and it is an
array, it will be appended to. Otherwise, the "tag" field will be left as is.

This setting can not be greater than the `spool max bytes` setting.

### `"persist directory"`

*String. Optional. Default: "."*  
*Requires restart*

The directory that Log Courier should store its persistence data in. The default
is the current working directory of Log Courier which is specified using the
path, `"."`.

At the time of writing, the only file saved here is `.log-courier` that
contains the offset in the file that Log Courier needs to resume from after a
graceful restart or server crash. The offset is only updated when the remote
server acknowledges receipt of the events.

### `"prospect interval"`

*Duration. Optional. Default: 10*

How often Log Courier should check for changes on the filesystem, such as the
appearance of new log files, rotations and deletions.

### `"spool max bytes"`

*Number. Optional. Default: 10485760*

The maximum size of an event spool, before compression. If an incomplete spool
does not have enough room for the next event, it will be flushed immediately.

If this value is modified, the receiving end should also be configured with the
new limit. For the Logstash plugin, this is the `max_packet_size` setting.

The maximum value for this setting is 2147483648 (2 GiB).

### `"spool size"`

*Number. Optional. Default: 1024*

How many events to spool together and flush at once. This improves efficiency
when processing large numbers of events by submitting them for processing in
bulk.

Internal benchmarks have shown that increasing to 5120, for example, can
give around a 25% boost of events per second, at the expense of more memory
usage.

*For most installations you should leave this at the default as it can
easily cope with over 10,000 events a second and uses little memory. It is
useful only in very specific circumstances.*

### `"spool timeout"`

*Duration. Optional. Default: 5*

The maximum amount of time to wait for a full spool. If an incomplete spool is
not filled within this time limit, the spool will be flushed immediately.

## `"network"`

The network configuration tells Log Courier where to ship the logs, and also
what transport and security to use.

### `"max pending payloads"`

*Number. Optional. Default: 4*

The maximum number of spools that can be in transit to a single server at any
one time. Each spool will be kept in memory until the remote server acknowledges
it.

If Log Courier has sent this many spools to a remote server, and has not yet
received acknowledgement responses for them (either because the remote server is
busy or because the link has high latency), it will pause and wait before
sending anymore.

*For most installations you should leave this at the default as it is high
enough to maintain throughput even on high latency links and low enough not to
cause excessive memory usage.*

### `"method"`

*String. Optional. Default: "failover"
Available values: "failover", "loadbalance"*

Specified the method to use when managing multiple `"servers"`.

`"failover"`: Connect to all servers, preferring the first server in the list
that has not failed. If the preferred server fails, gracefully failover to the
next available server. When a server recovers that is more preferred than the
current server, failback gracefully.

`"loadbalance"`: Connect to all servers, load balancing events between them.
Faster servers will receive more events than slower servers. Equally fast
servers will receive events in a round robin fashion.

### `"reconnect"`

*Duration. Optional. Default: 1*

Pause this long before reconnecting to a server. If the remote server is
completely down, this slows down the rate of reconnection attempts.

### `"rfc 2782 srv"`

*Boolean. Optional. Default: true*

When performing SRV DNS lookups for entries in the [`"servers"`](#servers) list,
use RFC 2782 style lookups of the form `_service._proto.example.com`.

### `"rfc 2782 service"`

*String. Optional. Default: "courier"*

Specifies the service to request when using RFC 2782 style SRV lookups. Using
the default, "courier", an "@example.com" server entry would result in a lookup
for `_courier._tcp.example.com`.

### `"servers"`

*Array of Strings. Required*

Sets the list of servers to send logs to. Accepted formats for each server entry
are:

* `ipaddress:port`
* `hostname:port` (A DNS lookup is performed)
* `@hostname` (A SRV DNS lookup is performed, with further DNS lookups if
required)

The initial server is randomly selected. Subsequent connection attempts are made
to the next IP address available (if the server had multiple IP addresses) or to
the next server listed in the configuration file (if all addresses for the
previous server were exausted.)

### `"ssl ca"`

*Filepath. Required with "transport" = "tls". Not allowed otherwise*

Path to a PEM encoded certificate file to use to verify the connected server.

### `"ssl certificate"`

*Filepath. Optional with "transport" = "tls". Not allowed otherwise*

Path to a PEM encoded certificate file to use as the client certificate.

### `"ssl key"`

*Filepath. Required with "ssl certificate". Not allowed when "transport" !=
"tls"*

Path to a PEM encoded private key to use with the client certificate.

### `"timeout"`

*Duration. Optional. Default: 15*

This is the maximum time Log Courier will wait for a server to respond to a
request after logs were send to it. If the server does not respond within this
time period the connection will be closed and reset.

### `"transport"`

*String. Optional. Default: "tls"  
Available values: "tcp", "tls"*

<!-- *Depending on how log-courier was built, some transports may not be available.
Run `log-courier -list-supported` to see the list of transports available in
a specific build of log-courier.* -->

Sets the transport to use when sending logs to the servers. "tls" is recommended
for most users.

"tcp" is an **insecure** equivalent to "tls" that does not encrypt traffic or
authenticate the identity of servers. This should only be used on trusted
internal networks. If in doubt, use the secure authenticating transport "tls".

## `"files"`

The files configuration lists the file groups that contain the logs you wish to
ship. It is an array of file group configurations.

```
    [
        {
            # First file group
        },
        {
            # Second file group
        }
    ]
```


src/lc-lib/config/config.go
/*
 * Copyright 2014 Jason Woods.
 *
 * This file is a modification of code from Logstash Forwarder.
 * Copyright 2012-2013 Jordan Sissel and contributors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package config

import (
	"bytes"
	"encoding/json"
	"fmt"
	"github.com/op/go-logging"
	"math"
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"time"
)

const (
	default_GeneralConfig_AdminEnabled       bool          = false
	default_GeneralConfig_AdminBind          string        = "tcp:127.0.0.1:1234"
	default_GeneralConfig_PersistDir         string        = "."
	default_GeneralConfig_ProspectInterval   time.Duration = 10 * time.Second
	default_GeneralConfig_SpoolSize          int64         = 1024
	default_GeneralConfig_SpoolMaxBytes      int64         = 10485760
	default_GeneralConfig_SpoolTimeout       time.Duration = 5 * time.Second
	default_GeneralConfig_LineBufferBytes    int64         = 16384
	default_GeneralConfig_MaxLineBytes       int64         = 1048576
	default_GeneralConfig_LogLevel           logging.Level = logging.INFO
	default_GeneralConfig_LogStdout          bool          = true
	default_GeneralConfig_LogSyslog          bool          = false
	default_NetworkConfig_Method             string        = "failover"
	default_NetworkConfig_Transport          string        = "tls"
	default_NetworkConfig_Rfc2782Srv         bool          = true
	default_NetworkConfig_Rfc2782Service     string        = "courier"
	default_NetworkConfig_Timeout            time.Duration = 15 * time.Second
	default_NetworkConfig_Reconnect          time.Duration = 1 * time.Second
	default_NetworkConfig_MaxPendingPayloads int64         = 4
	default_StreamConfig_Codec               string        = "plain"
	default_StreamConfig_DeadTime            int64         = 86400
)

var (
	default_GeneralConfig_Host string = "localhost.localdomain"
)

type General struct {
	AdminEnabled     bool          `config:"admin enabled"`
	AdminBind        string        `config:"admin listen address"`
	PersistDir       string        `config:"persist directory"`
	ProspectInterval time.Duration `config:"prospect interval"`
	SpoolSize        int64         `config:"spool size"`
	SpoolMaxBytes    int64         `config:"spool max bytes"`
	SpoolTimeout     time.Duration `config:"spool timeout"`
	LineBufferBytes  int64         `config:"line buffer bytes"`
	MaxLineBytes     int64         `config:"max line bytes"`
	LogLevel         logging.Level `config:"log level"`
	LogStdout        bool          `config:"log stdout"`
	LogSyslog        bool          `config:"log syslog"`
	LogFile          string        `config:"log file"`
	Host             string        `config:"host"`
	GlobalFields     map[string]interface{} `config:"global fields"`
}

type Network struct {
	Transport          string        `config:"transport"`
	Servers            []string      `config:"servers"`
	Method             string        `config:"method"`
	Rfc2782Srv         bool          `config:"rfc 2782 srv"`
	Rfc2782Service     string        `config:"rfc 2782 service"`
	Timeout            time.Duration `config:"timeout"`
	Reconnect          time.Duration `config:"reconnect"`
	MaxPendingPayloads int64         `config:"max pending payloads"`
	Factory            interface{}
	Unused             map[string]interface{}
}

type CodecStub struct {
	Name    string `config:"name"`
	Unused  map[string]interface{}
	Factory interface{}
}

type Stream struct {
	Fields   map[string]interface{} `config:"fields"`
	Codecs   []CodecStub            `config:"codecs"`
	DeadTime time.Duration          `config:"dead time"`
}

type File struct {
	Paths  []string `config:"paths"`
	Stream          `config:",embed"`
}

type Config struct {
	General  General  `config:"general"`
	Network  Network  `config:"network"`
	Files    []File   `config:"files"`
	Includes []string `config:"includes"`
	Stdin    Stream   `config:"stdin"`
}

func NewConfig() *Config {
	return &Config{}
}

func (c *Config) loadFile(path string) (stripped *bytes.Buffer, err error) {
	stripped = new(bytes.Buffer)

	file, err := os.Open(path)
	if err != nil {
		err = fmt.Errorf("Failed to open config file: %s", err)
		return
	}
	defer file.Close()

	stat, err := file.Stat()
	if err != nil {
		err = fmt.Errorf("Stat failed for config file: %s", err)
		return
	}
	if stat.Size() == 0 {
		err = fmt.Errorf("Empty configuration file")
		return
	}
	if stat.Size() > (10 << 20) {
		err = fmt.Errorf("Config file too large (%s)", stat.Size())
		return
	}

	// Strip comments and read config into stripped
	var s, p, state int
	{
		// Pull the config file into memory
		buffer := make([]byte, stat.Size())
		_, err = file.Read(buffer)
		if err != nil {
			return
		}

		for p < len(buffer) {
			b := buffer[p]
			if state == 0 {
				// Main body
				if b == '"' {
					state = 1
				} else if b == '\'' {
					state = 2
				} else if b == '#' {
					state = 3
					stripped.Write(buffer[s:p])
				} else if b == '/' {
					state = 4
				}
			} else if state == 1 {
				// Double-quoted string
				if b == '\\' {
					state = 5
				} else if b == '"' {
					state = 0
				}
			} else if state == 2 {
				// Single-quoted string
				if b == '\\' {
					state = 6
				} else if b == '\'' {
					state = 0
				}
			} else if state == 3 {
				// End of line comment (#)
				if b == '\r' || b == '\n' {
					state = 0
					s = p + 1
				}
			} else if state == 4 {
				// Potential start of multiline comment
				if b == '*' {
					state = 7
					stripped.Write(buffer[s : p-1])
				} else {
					state = 0
				}
			} else if state == 5 {
				// Escape within double quote
				state = 1
			} else if state == 6 {
				// Escape within single quote
				state = 2
			} else if state == 7 {
				// Multiline comment (/**/)
				if b == '*' {
					state = 8
				}
			} else { // state == 8
				// Potential end of multiline comment
				if b == '/' {
					state = 0
					s = p + 1
				} else {
					state = 7
				}
			}
			p++
		}
		stripped.Write(buffer[s:p])
	}

	if stripped.Len() == 0 {
		err = fmt.Errorf("Empty configuration file")
		return
	}

	return
}

// Parse a *json.SyntaxError into a pretty error message
func (c *Config) parseSyntaxError(js []byte, err error) error {
	json_err, ok := err.(*json.SyntaxError)
	if !ok {
		return err
	}

	start := bytes.LastIndex(js[:json_err.Offset], []byte("\n"))+1
	end := bytes.Index(js[start:], []byte("\n"))
	if end >= 0 {
		end += start
	} else {
		end = len(js)
	}

	line, pos := bytes.Count(js[:start], []byte("\n")), int(json_err.Offset) - start - 1

	var posStr string
	if pos > 0 {
		posStr = strings.Repeat(" ", pos)
	} else {
		posStr = ""
	}

	return fmt.Errorf("%s on line %d\n%s\n%s^", err, line, js[start:end], posStr)
}

// TODO: Config from a TOML? Maybe a custom one
func (c *Config) Load(path string) (err error) {
	var data *bytes.Buffer

	// Read the main config file
	if data, err = c.loadFile(path); err != nil {
		return
	}

	// Pull the entire structure into raw_config
	raw_config := make(map[string]interface{})
	if err = json.Unmarshal(data.Bytes(), &raw_config); err != nil {
		return c.parseSyntaxError(data.Bytes(), err)
	}

	// Fill in defaults where the zero-value is a valid setting
	c.General.AdminEnabled = default_GeneralConfig_AdminEnabled
	c.General.LogLevel = default_GeneralConfig_LogLevel
	c.General.LogStdout = default_GeneralConfig_LogStdout
	c.General.LogSyslog = default_GeneralConfig_LogSyslog
	c.Network.Rfc2782Srv = default_NetworkConfig_Rfc2782Srv

	// Populate configuration - reporting errors on spelling mistakes etc.
	if err = c.PopulateConfig(c, "/", raw_config); err != nil {
		return
	}

	// Iterate includes
	for _, glob := range c.Includes {
		// Glob the path
		var matches []string
		if matches, err = filepath.Glob(glob); err != nil {
			return
		}

		for _, include := range matches {
			// Read the include
			if data, err = c.loadFile(include); err != nil {
				return
			}

			// Pull the structure into raw_include
			raw_include := make([]interface{}, 0)
			if err = json.Unmarshal(data.Bytes(), &raw_include); err != nil {
				return
			}

			// Append to configuration
			if err = c.populateConfigSlice(reflect.ValueOf(c).Elem().FieldByName("Files"), fmt.Sprintf("%s/", include), raw_include); err != nil {
				return
			}
		}
	}

	if c.General.AdminBind == "" {
		c.General.AdminBind = default_GeneralConfig_AdminBind
	}

	if c.General.PersistDir == "" {
		c.General.PersistDir = default_GeneralConfig_PersistDir
	}

	if c.General.ProspectInterval == time.Duration(0) {
		c.General.ProspectInterval = default_GeneralConfig_ProspectInterval
	}

	if c.General.SpoolSize == 0 {
		c.General.SpoolSize = default_GeneralConfig_SpoolSize
	}

	// Enforce maximum of 2 GB since event transmit length is uint32
	if c.General.SpoolMaxBytes == 0 {
		c.General.SpoolMaxBytes = default_GeneralConfig_SpoolMaxBytes
	}
	if c.General.SpoolMaxBytes > 2*1024*1024*1024 {
		err = fmt.Errorf("/general/spool max bytes can not be greater than 2 GiB")
		return
	}

	if c.General.SpoolTimeout == time.Duration(0) {
		c.General.SpoolTimeout = default_GeneralConfig_SpoolTimeout
	}

	if c.General.LineBufferBytes == 0 {
		c.General.LineBufferBytes = default_GeneralConfig_LineBufferBytes
	}
	if c.General.LineBufferBytes < 1 {
		err = fmt.Errorf("/general/line buffer bytes must be greater than 1")
		return
	}

	// Max line bytes can not be larger than spool max bytes
	if c.General.MaxLineBytes == 0 {
		c.General.MaxLineBytes = default_GeneralConfig_MaxLineBytes
	}
	if c.General.MaxLineBytes > c.General.SpoolMaxBytes {
		err = fmt.Errorf("/general/max line bytes can not be greater than /general/spool max bytes")
		return
	}

	if c.General.Host == "" {
		ret, err := os.Hostname()
		if err == nil {
			c.General.Host = ret
		} else {
			c.General.Host = default_GeneralConfig_Host
			log.Warning("Failed to determine the FQDN; using '%s'.", c.General.Host)
		}
	}

	if c.Network.Method == "" {
		c.Network.Method = default_NetworkConfig_Method
	}
	if c.Network.Method != "failover" && c.Network.Method != "loadbalance" {
		err = fmt.Errorf("The network method (/network/method) is not recognised: %s", c.Network.Method)
		return
	}

	if len(c.Network.Servers) == 0 {
		err = fmt.Errorf("No network servers were specified (/network/servers)")
		return
	}

	servers := make(map[string]bool)
	for _, server := range c.Network.Servers {
		if _, exists := servers[server]; exists {
			err = fmt.Errorf("The list of network servers (/network/servers) must be unique: %s appears multiple times", server)
			return
		}
		servers[server] = true
	}
	servers = nil

	if c.Network.Transport == "" {
		c.Network.Transport = default_NetworkConfig_Transport
	}

	if registrarFunc, ok := registeredTransports[c.Network.Transport]; ok {
		if c.Network.Factory, err = registrarFunc(c, "/network/", c.Network.Unused, c.Network.Transport); err != nil {
			return
		}
	} else {
		err = fmt.Errorf("Unrecognised transport '%s'", c.Network.Transport)
		return
	}

	if c.Network.Rfc2782Service == "" {
		c.Network.Rfc2782Service = default_NetworkConfig_Rfc2782Service
	}
	if c.Network.Timeout == time.Duration(0) {
		c.Network.Timeout = default_NetworkConfig_Timeout
	}
	if c.Network.Reconnect == time.Duration(0) {
		c.Network.Reconnect = default_NetworkConfig_Reconnect
	}
	if c.Network.MaxPendingPayloads == 0 {
		c.Network.MaxPendingPayloads = default_NetworkConfig_MaxPendingPayloads
	}

	for k := range c.Files {
		if len(c.Files[k].Paths) == 0 {
			err = fmt.Errorf("No paths specified for /files[%d]/", k)
			return
		}

		if err = c.initStreamConfig(fmt.Sprintf("/files[%d]", k), &c.Files[k].Stream); err != nil {
			return
		}
	}

	if err = c.initStreamConfig("/stdin", &c.Stdin); err != nil {
		return
	}

	return
}

func (c *Config) initStreamConfig(path string, stream_config *Stream) (err error) {
	if len(stream_config.Codecs) == 0 {
		stream_config.Codecs = []CodecStub{CodecStub{Name: default_StreamConfig_Codec}}
	}

	for i := 0; i < len(stream_config.Codecs); i++ {
		codec := &stream_config.Codecs[i]
		if registrarFunc, ok := registeredCodecs[codec.Name]; ok {
			if codec.Factory, err = registrarFunc(c, path, codec.Unused, codec.Name); err != nil {
				return
			}
		} else {
			return fmt.Errorf("Unrecognised codec '%s' for %s", codec.Name, path)
		}
	}

	if stream_config.DeadTime == time.Duration(0) {
		stream_config.DeadTime = time.Duration(default_StreamConfig_DeadTime) * time.Second
	}

	// TODO: EDGE CASE: Event transmit length is uint32, if fields length is rediculous we will fail

	return nil
}

// PopulateConfig populates dynamic configuration, automatically converting time.Duration etc.
// Any config entries not found in the structure are moved to an "Unused" field if it exists
// or an error is reported if "Unused" is not available
// We can then take the unused configuration dynamically at runtime based on another value
func (c *Config) PopulateConfig(config interface{}, config_path string, raw_config map[string]interface{}) (err error) {
	vconfig := reflect.ValueOf(config).Elem()
FieldLoop:
	for i := 0; i < vconfig.NumField(); i++ {
		field := vconfig.Field(i)
		if !field.CanSet() {
			continue
		}
		fieldtype := vconfig.Type().Field(i)
		tag := fieldtype.Tag.Get("config")
		mods := strings.Split(tag, ",")
		tag = mods[0]
		mods = mods[1:]
		for _, mod := range mods {
			if mod == "embed" && field.Kind() == reflect.Struct {
				if err = c.PopulateConfig(field.Addr().Interface(), config_path, raw_config); err != nil {
					return
				}
				continue FieldLoop
			}
		}
		if tag == "" {
			continue
		}
		if _, ok := raw_config[tag]; ok {
			if field.Kind() == reflect.Struct {
				if reflect.TypeOf(raw_config[tag]).Kind() == reflect.Map {
					if err = c.PopulateConfig(field.Addr().Interface(), fmt.Sprintf("%s%s/", config_path, tag), raw_config[tag].(map[string]interface{})); err != nil {
						return
					}
					delete(raw_config, tag)
					continue
				} else {
					err = fmt.Errorf("Option %s%s must be a hash", config_path, tag)
					return
				}
			}
			value := reflect.ValueOf(raw_config[tag])
			if value.Type().AssignableTo(field.Type()) {
				field.Set(value)
			} else if value.Kind() == reflect.Slice && field.Kind() == reflect.Slice {
				if err = c.populateConfigSlice(field, fmt.Sprintf("%s%s/", config_path, tag), raw_config[tag].([]interface{})); err != nil {
					return
				}
			} else if value.Kind() == reflect.Map && field.Kind() == reflect.Map {
				if field.IsNil() {
					field.Set(reflect.MakeMap(field.Type()))
				}
				for _, j := range value.MapKeys() {
					item := value.MapIndex(j)
					if item.Elem().Type().AssignableTo(field.Type().Elem()) {
						field.SetMapIndex(j, item.Elem())
					} else {
						err = fmt.Errorf("Option %s%s[%s] must be %s or similar", config_path, tag, j, field.Type().Elem())
						return
					}
				}
			} else if field.Type().String() == "time.Duration" {
				var duration float64
				vduration := reflect.ValueOf(duration)
				fail := true
				if value.Type().AssignableTo(vduration.Type()) {
					duration = value.Float()
					if duration >= math.MinInt64 && duration <= math.MaxInt64 {
						field.Set(reflect.ValueOf(time.Duration(int64(duration)) * time.Second))
						fail = false
					}
				} else if value.Kind() == reflect.String {
					var tduration time.Duration
					if tduration, err = time.ParseDuration(value.String()); err == nil {
						field.Set(reflect.ValueOf(tduration))
						fail = false
					}
				}
				if fail {
					err = fmt.Errorf("Option %s%s must be a valid numeric or string duration", config_path, tag)
					return
				}
			} else if field.Type().String() == "logging.Level" {
				fail := true
				if value.Kind() == reflect.String {
					var llevel logging.Level
					if llevel, err = logging.LogLevel(value.String()); err == nil {
						fail = false
						field.Set(reflect.ValueOf(llevel))
					}
				}
				if fail {
					err = fmt.Errorf("Option %s%s is not a valid log level (critical, error, warning, notice, info, debug)", config_path, tag)
					return
				}
			} else if field.Kind() == reflect.Int64 {
				fail := true
				if value.Kind() == reflect.Float64 {
					number := value.Float()
					if math.Floor(number) == number {
						fail = false
						field.Set(reflect.ValueOf(int64(number)))
					}
				}
				if fail {
					err = fmt.Errorf("Option %s%s is not a valid integer", config_path, tag, field.Type())
					return
				}
			} else if field.Kind() == reflect.Int {
				fail := true
				if value.Kind() == reflect.Float64 {
					number := value.Float()
					if math.Floor(number) == number {
						fail = false
						field.Set(reflect.ValueOf(int(number)))
					}
				}
				if fail {
					err = fmt.Errorf("Option %s%s is not a valid integer", config_path, tag, field.Type())
					return
				}
			} else {
				err = fmt.Errorf("Option %s%s must be %s or similar (%s provided)", config_path, tag, field.Type(), value.Type())
				return
			}
			delete(raw_config, tag)
		}
	}
	if unused := vconfig.FieldByName("Unused"); unused.IsValid() {
		if unused.IsNil() {
			unused.Set(reflect.MakeMap(unused.Type()))
		}
		for k, v := range raw_config {
			unused.SetMapIndex(reflect.ValueOf(k), reflect.ValueOf(v))
		}
		return
	}
	return c.ReportUnusedConfig(config_path, raw_config)
}

func (c *Config) populateConfigSlice(field reflect.Value, config_path string, raw_config []interface{}) (err error) {
	elemtype := field.Type().Elem()
	if elemtype.Kind() == reflect.Struct {
		for j := 0; j < len(raw_config); j++ {
			item := reflect.New(elemtype)
			if err = c.PopulateConfig(item.Interface(), fmt.Sprintf("%s[%d]/", config_path, j), raw_config[j].(map[string]interface{})); err != nil {
				return
			}
			field.Set(reflect.Append(field, item.Elem()))
		}
	} else {
		for j := 0; j < len(raw_config); j++ {
			field.Set(reflect.Append(field, reflect.ValueOf(raw_config[j])))
		}
	}
	return
}

func (c *Config) ReportUnusedConfig(config_path string, raw_config map[string]interface{}) (err error) {
	for k := range raw_config {
		err = fmt.Errorf("Option %s%s is not available", config_path, k)
		return
	}
	return
}
